In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score

The last column of this dataset denotes whether the e-mail was 
considered spam (1) or not (0).  
Most of the attributes indicate whether a particular word or
character was frequently occuring in the e-mail

We will be using this dataset to train a model to make predictions if an email is consided spam or not when presented with these atributes.

# Dataset Classification

In [2]:
df=pd.read_csv('spambase.data')
df.head()

0  0.64  0.64.1  0.1  0.32   0.2   0.3   0.4   0.5   0.6  ...  0.41  \
0  0.21  0.28    0.50  0.0  0.14  0.28  0.21  0.07  0.00  0.94  ...  0.00   
1  0.06  0.00    0.71  0.0  1.23  0.19  0.19  0.12  0.64  0.25  ...  0.01   
2  0.00  0.00    0.00  0.0  0.63  0.00  0.31  0.63  0.31  0.63  ...  0.00   
3  0.00  0.00    0.00  0.0  0.63  0.00  0.31  0.63  0.31  0.63  ...  0.00   
4  0.00  0.00    0.00  0.0  1.85  0.00  0.00  1.85  0.00  0.00  ...  0.00   

    0.42  0.43  0.778   0.44   0.45  3.756   61   278  1  
0  0.132   0.0  0.372  0.180  0.048  5.114  101  1028  1  
1  0.143   0.0  0.276  0.184  0.010  9.821  485  2259  1  
2  0.137   0.0  0.137  0.000  0.000  3.537   40   191  1  
3  0.135   0.0  0.135  0.000  0.000  3.537   40   191  1  
4  0.223   0.0  0.000  0.000  0.000  3.000   15    54  1  

[5 rows x 58 columns]

In [3]:
#defining x as input and y output split
X=df.iloc[:,0:57]
Y=df.iloc[:,57]


In [4]:
#Spliting into train and test sets using train_test_split
X_train,X_test,y_train,y_test=train_test_split(X,Y,test_size=0.3,random_state=3)

In [5]:
print(X_train.shape)
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)

(3220, 57)
(1380, 57)
(3220,)
(1380,)


In [6]:
# Feature Scaling
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Actual Model

In [7]:
model=LogisticRegression()

In [8]:
model.fit(X_train_scaled,y_train)

LogisticRegression()

# Model evaluation

In [9]:
train_predict=model.predict(X_train_scaled)
training_accuracy=accuracy_score(y_train,train_predict)

In [10]:
test_predict=model.predict(X_test_scaled)
test_accuracy=accuracy_score(y_test,test_predict)

In [29]:
#Printing model accuracy
print(f'model accuracy on training data= {round(training_accuracy*100)}%')
print(f'model accuracy on test data= {round(test_accuracy*100)}%')

model accuracy on training data= 93%
model accuracy on test data= 93%


In [26]:
#Comparing model predictions to actual value in dataset
Spam_predict=np.where(test_predict==1,'Spam',"Not Spam") #Classifying predictions as spam or not spam
Spam_actual=Spam_predict=np.where(y_test==1,'Spam',"Not Spam")
indexes=np.random.choice(len(Spam_predict),10,replace=False) #Geting 10 random indexes so loop is diffrent everytime

for i in indexes:
    print(f'Model prediction= {Spam_predict[i]:<10}  :  Actual value in dataset= {Spam_actual[i]:<10}')

Model prediction= Not Spam    :  Actual value in dataset= Not Spam  
Model prediction= Not Spam    :  Actual value in dataset= Not Spam  
Model prediction= Not Spam    :  Actual value in dataset= Not Spam  
Model prediction= Not Spam    :  Actual value in dataset= Not Spam  
Model prediction= Not Spam    :  Actual value in dataset= Not Spam  
Model prediction= Spam        :  Actual value in dataset= Spam      
Model prediction= Not Spam    :  Actual value in dataset= Not Spam  
Model prediction= Spam        :  Actual value in dataset= Spam      
Model prediction= Spam        :  Actual value in dataset= Spam      
Model prediction= Spam        :  Actual value in dataset= Spam      


# Final Acuracy of model

**Model acuracy for training data=** 93%

**Model acuracy for test data=** 93%